src/item.cpp

```diff
@@ -4341,68 +4341,92 @@ item_location item::pick_reload_ammo( player &u, bool interactive ) const
+    };
+
+    // Pad the first column including 4 trailing spaces
+    int w = pad( names, utf8_width( menu.text ), 6 );
+    menu.text.insert( 0, 2, ' ' ); // add space for UI hotkeys
+    menu.text += std::string( w + 2 - utf8_width( menu.text ), ' ' );
+    menu.w_width += w + 2;
+
+    // Pad the location similarly (excludes leading "| " and trailing " ")
+    w = pad( where, utf8_width( _( "| Location " ) ) - 3, 6 );
+    menu.text += _("| Location " );
+    menu.text += std::string( w + 3 - utf8_width( _( "| Location " ) ), ' ' );
+    menu.w_width += w + 3;
+
+    // We only show ammo statistics for guns
+    if( is_gun() ) {
+        menu.text += _( "| Damage  | Pierce  | Range   | Accuracy" );
+        menu.w_width += 40;
+    }
+
+    menu.w_width += 6; // include space for borders
+
+    // center dialog
+    menu.w_y = std::max( ( TERMX / 2 ) - int( menu.w_width / 2 ) , 0 );
+    menu.w_y = std::max( ( TERMY / 2 ) - int( (ammo_list.size() + 3 ) / 2 ) , 0 );
 
     std::string lastreload = "";
     if( uistate.lastreload.find( ammo_type() ) != uistate.lastreload.end() ) {
         lastreload = uistate.lastreload[ ammo_type() ];
     }
 
-    amenu.text = std::string( _( "Choose ammo type:" ) );
-    if( utf8_width(amenu.text) < namelen ) {
-        amenu.text += std::string( namelen - utf8_width(amenu.text), ' ' );
-    } else {
-        utf8_truncate( amenu.text, utf8_width(amenu.text) - namelen );
-    }
-    // To cover the space in the header that is used by the hotkeys created by uimenu
-    amenu.text.insert( 0, "  " );
-    //~ header of table that appears when reloading
-    amenu.text += _( "| Location                              | Damage  | Pierce  | Range   | Accuracy" );
-    int i = 0;
-    for( auto& e : ammo_list ) {
-        const item *it = e.get_item();
-        const auto curammo = it->ammo_data(); // nullptr for empty magazines
+    for( auto i = 0; i != (int) ammo_list.size(); ++i ) {
+        const item *it = ammo_list[i].get_item();
+        std::string row = names[i] + "| " + where[i] + " ";
 
-        std::string row;
-        if( it->is_magazine() && curammo ) {
-            long qty = it->ammo_remaining();
-            row = string_format( "%s with %s (%d)", it->type->nname( 1 ).c_str(), curammo->nname( qty ).c_str(), qty );
-        } else {
-            row = it->display_name();
-        }
-        if( utf8_width(row) < namelen ) {
-            row += std::string( namelen - utf8_width(row), ' ' );
-        } else {
-            utf8_truncate( row, utf8_width(row) - namelen );
-        }
-
-        const auto ammo_damage     = curammo ? curammo->ammo->damage : 0;
-        const auto ammo_pierce     = curammo ? curammo->ammo->pierce : 0;
-        const auto ammo_range      = curammo ? curammo->ammo->range  : 0;
-        const auto ammo_dispersion = curammo ? 100 - curammo->ammo->dispersion : 0;
+        if( is_gun() ) {
+            const itype *curammo = ammo_data(); // nullptr for empty magazines
```

Comment:
This looks like you're getting and consequently displaying the ammo data of the gun item itself. It will be the same for all entries in the menu.